File: adapters/bank-adapter/java/src/main/java/com/elfort/efn/adapter/EFNAdapterController.java
package com.elfort.efn.adapter;

import com.elfort.efn.adapter.models.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import jakarta.servlet.http.HttpServletRequest;
import java.nio.charset.StandardCharsets;
import java.util.Map;

@RestController
@RequestMapping("/efn/v1")
public class EFNAdapterController {

    private final EFNBankAdapter adapter;
    private final String apiSecret;
    private final ObjectMapper mapper = new ObjectMapper();

    public EFNAdapterController(EFNBankAdapter adapter, String apiSecret) {
        this.adapter = adapter;
        this.apiSecret = apiSecret;
    }

    private ResponseEntity<?> sigError() {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
            .body(Map.of("success", false, "message", "Invalid or expired signature"));
    }

    private byte[] readBody(HttpServletRequest req) {
        try { return req.getInputStream().readAllBytes(); } catch (Exception e) { return new byte[0]; }
    }

    private boolean checkSig(HttpServletRequest req, byte[] body) {
        return EFNSignatureVerifier.verify(
            apiSecret,
            req.getHeader("X-EFN-Timestamp"),
            body,
            req.getHeader("X-EFN-Signature")
        );
    }

    @PostMapping("/authorizations")
    public ResponseEntity<?> authorize(HttpServletRequest req) throws Exception {
        byte[] body = readBody(req);
        if (!checkSig(req, body)) return sigError();
        return ResponseEntity.ok(adapter.authorize(mapper.readValue(body, AuthorizationRequest.class)));
    }

    @PostMapping("/authorizations/{id}/capture")
    public ResponseEntity<?> capture(@PathVariable String id, HttpServletRequest req) throws Exception {
        byte[] body = readBody(req);
        if (!checkSig(req, body)) return sigError();
        return ResponseEntity.ok(adapter.capture(id, mapper.readValue(body, CaptureRequest.class)));
    }

    @PostMapping("/authorizations/{id}/reversal")
    public ResponseEntity<?> reversal(@PathVariable String id, HttpServletRequest req) throws Exception {
        byte[] body = readBody(req);
        if (!checkSig(req, body)) return sigError();
        return ResponseEntity.ok(adapter.reverse(id, mapper.readValue(body, ReversalRequest.class)));
    }

    @PostMapping("/debit")
    public ResponseEntity<?> debit(HttpServletRequest req) throws Exception {
        byte[] body = readBody(req);
        if (!checkSig(req, body)) return sigError();
        return ResponseEntity.ok(adapter.debit(mapper.readValue(body, DebitRequest.class)));
    }

    @PostMapping("/credit")
    public ResponseEntity<?> credit(HttpServletRequest req) throws Exception {
        byte[] body = readBody(req);
        if (!checkSig(req, body)) return sigError();
        return ResponseEntity.ok(adapter.credit(mapper.readValue(body, DebitRequest.class)));
    }

    @PostMapping("/balance")
    public ResponseEntity<?> balance(HttpServletRequest req) throws Exception {
        byte[] body = readBody(req);
        if (!checkSig(req, body)) return sigError();
        return ResponseEntity.ok(adapter.balance(mapper.readValue(body, BalanceRequest.class)));
    }

    @PostMapping("/account-enquiry")
    public ResponseEntity<?> accountEnquiry(HttpServletRequest req) throws Exception {
        byte[] body = readBody(req);
        if (!checkSig(req, body)) return sigError();
        return ResponseEntity.ok(adapter.accountEnquiry(mapper.readValue(body, AccountEnquiryRequest.class)));
    }

    @GetMapping("/transaction/{ref}/status")
    public ResponseEntity<?> txStatus(@PathVariable String ref) {
        return ResponseEntity.ok(adapter.transactionStatus(ref));
    }

    @PostMapping("/consent-otp")
    public ResponseEntity<?> consentOtp(HttpServletRequest req) throws Exception {
        byte[] body = readBody(req);
        if (!checkSig(req, body)) return sigError();
        return ResponseEntity.ok(adapter.consentOtp(mapper.readValue(body, ConsentOTPRequest.class)));
    }

    @PostMapping("/consent-verify")
    public ResponseEntity<?> consentVerify(HttpServletRequest req) throws Exception {
        byte[] body = readBody(req);
        if (!checkSig(req, body)) return sigError();
        return ResponseEntity.ok(adapter.consentVerify(mapper.readValue(body, ConsentVerifyRequest.class)));
    }

    @GetMapping("/health")
    public ResponseEntity<?> health() {
        return ResponseEntity.ok(Map.of("status", "ok"));
    }
}

File: csharp/EfnAdapter/EfnAdapterController.cs
using Microsoft.AspNetCore.Mvc;

namespace EfnAdapter;

[ApiController]
public class EfnAdapterController(IEfnBankAdapter adapter, EfnAdapterOptions options) : ControllerBase
{
    private bool Verify()
    {
        Request.Headers.TryGetValue("X-EFN-Timestamp", out var ts);
        Request.Headers.TryGetValue("X-EFN-Signature", out var sig);
        // Body already read via EnableBuffering middleware
        Request.Body.Position = 0;
        using var ms = new System.IO.MemoryStream();
        Request.Body.CopyTo(ms);
        return EfnSecurity.VerifySignature(options.ApiSecret, ts!, ms.ToArray(), sig!);
    }

    [HttpPost("/efn/v1/authorizations")]
    public async Task<IActionResult> Authorize([FromBody] AuthorizationRequest req)
    {
        if (!Verify()) return Unauthorized(new { success = false, message = "Invalid signature" });
        return Ok(await adapter.AuthorizeAsync(req));
    }

    [HttpPost("/efn/v1/authorizations/{authorizationId}/capture")]
    public async Task<IActionResult> Capture(string authorizationId, [FromBody] AuthorizationRequest req)
    {
        if (!Verify()) return Unauthorized(new { success = false, message = "Invalid signature" });
        return Ok(await adapter.CaptureAsync(authorizationId, req));
    }

    [HttpPost("/efn/v1/authorizations/{authorizationId}/reversal")]
    public async Task<IActionResult> Reversal(string authorizationId, [FromBody] AuthorizationRequest req)
    {
        if (!Verify()) return Unauthorized(new { success = false, message = "Invalid signature" });
        return Ok(await adapter.ReverseAsync(authorizationId, req));
    }

    [HttpPost("/efn/v1/debit")]
    public async Task<IActionResult> Debit([FromBody] DebitRequest req)
    {
        if (!Verify()) return Unauthorized(new { success = false, message = "Invalid signature" });
        return Ok(await adapter.DebitAsync(req));
    }

    [HttpPost("/efn/v1/credit")]
    public async Task<IActionResult> Credit([FromBody] CreditRequest req)
    {
        if (!Verify()) return Unauthorized(new { success = false, message = "Invalid signature" });
        return Ok(await adapter.CreditAsync(req));
    }

    [HttpPost("/efn/v1/balance")]
    public async Task<IActionResult> Balance([FromBody] BalanceRequest req)
    {
        if (!Verify()) return Unauthorized(new { success = false, message = "Invalid signature" });
        return Ok(await adapter.BalanceAsync(req));
    }

    [HttpPost("/efn/v1/account-enquiry")]
    public async Task<IActionResult> AccountEnquiry([FromBody] AccountEnquiryRequest req)
    {
        if (!Verify()) return Unauthorized(new { success = false, message = "Invalid signature" });
        return Ok(await adapter.AccountEnquiryAsync(req));
    }

    [HttpGet("/efn/v1/transaction/{efnReference}/status")]
    public async Task<IActionResult> TransactionStatus(string efnReference)
        => Ok(await adapter.TransactionStatusAsync(efnReference));

    [HttpPost("/efn/v1/consent-otp")]
    public async Task<IActionResult> ConsentOtp([FromBody] ConsentOtpRequest req)
    {
        if (!Verify()) return Unauthorized(new { success = false, message = "Invalid signature" });
        return Ok(await adapter.ConsentOtpAsync(req));
    }

    [HttpPost("/efn/v1/consent-verify")]
    public async Task<IActionResult> ConsentVerify([FromBody] ConsentVerifyRequest req)
    {
        if (!Verify()) return Unauthorized(new { success = false, message = "Invalid signature" });
        return Ok(await adapter.ConsentVerifyAsync(req));
    }

    [HttpGet("/efn/v1/health")]
    public IActionResult Health() => Ok(new { status = "ok" });
}

public class EfnAdapterOptions { public string ApiSecret { get; set; } = ""; }
